packages/react-utilities/src/compose/types.ts

```diff
@@ -32,53 +34,86 @@ export type UnknownSlotProps = SlotProps<
  * those props.
  */
-export type SlotProps<Props extends { children?: React.ReactNode } = {}> = Props & {
+type WithSlotRenderFunction<Props extends { children?: React.ReactNode } = {}> = Props & {
   children?: Props['children'] | SlotRenderFunction<Props>;
 };
 
 /**
- * Defines the slot props for a slot that supports a Component type.
+ * The props type and shorthand value for a slot.
+ *
+ * The slot type can be either an intrinsic element like `'div'`, or a component like `typeof Button`.
  *
- * For intrinsic/native elements like 'div', use {@link IntrinsicSlotProps} instead.
+ * For intrinsic elements:
+ * * The first param is the slot's default type if no `as` prop is specified.
+ * * The second param is an optional union of alternative types that can be specified for the `as` prop.
  *
- * The generic param is the type of a control, i.e. a React component. For example:
+ * For components:
+ * * The first param is the component's type, like `typeof Button`
+ * * The second param is ignored and should not be specified
  *
  * @example
  * ```
- * type Props = {...}
- * const Button: React.FC<Props> = () => {...}
- * // $ExpectType ...
- * type SlotProps = ComponentSlotProps<typeof Button>
+ * // Intrinsic element examples:
+ * Slot<'div'> // Slot is always div
+ * Slot<'button', 'a'> // Defaults to button, but allows as="a" with anchor-specific props
+ * Slot<'label', 'span' | 'div'> // Defaults to label, but allows as="span" or as="div"
+ * NonNullable<Slot<'div'>> // Slot that will always be rendered and can't be set to null by the user
+ *
+ * // Component examples:
+ * Slot<typeof Button> // Slot is always a Button, and accepts all of Button's Props
+ * NonNullable<Slot<typeof Label>> // Slot is a Label and will always be rendered (can't be set to null by the user)
  * ```
  */
-export type ComponentSlotProps<Component extends React.ComponentType> = Component extends React.ComponentType<
-  infer Props
->
-  ? SlotProps<Props>
-  : never;
+export type Slot<
+  Type extends keyof JSX.IntrinsicElements | React.ComponentType,
+  AlternateAs extends keyof JSX.IntrinsicElements = never
+> =
+  | (Type extends keyof JSX.IntrinsicElements
+      ? IntrinsicSlotProps<Type, AlternateAs>
+      : Type extends React.ComponentType
+      ? ComponentSlotProps<Type>
+      : Type)
+  | SlotShorthandValue
+  | null;
 
 /**
- * Define the slot arguments for a slot that supports one or more intrinsic element types, such as 'div'.
- * For slots that support custom components, use {@link ComponentSlotProps} instead.
+ * Similar to {@link Slot}, but doesn't allow children to be specified through shorthand or props.
  *
- * The first param is the slot's default type if no `as` prop is specified.
- * The second param is an optional union of alternative types that can be specified for the `as` prop.
+ * This is intended to be used for element types that don't allow children, like `'input'` or `'img'`.
+ */
+export type SlotNoChildren<
```

Comment:
```suggestion
export type SlotWithoutChildren<
```

It's a personal preference, but `SlotWithoutChildren` sounds better IMO.

Reply:
What about reworking this into a utility type so you can do things like?:

```tsx
slot: WithoutChildren<Slot<'img'>>;
slot: NonNullable<WithoutChildren<Slot<typeof ChildlessCustomComponent>>>;
```

It's a bit wordy but does align with the pattern of using the `NonNullable` utility type.

Reply:
Throwing wood in the fire: ✨ _**ChildrenlessSlot**_ ✨

Reply:
Based on this comment https://github.com/microsoft/fluentui/pull/21518#discussion_r797406400, I'm going to remove `SlotNoChildren` and just make `Slot` do the right thing automatically.